WIP: Modernize CMake

Ram-Z opened the pull request:
I had an issue with the link interface of various pcl targets, which would include **ALL** VTK libraries, which would result in any targets using say `pcl_visualization` to also link against `libvtkIONetCDF.so.1` which is not required by it. The relevant changes are [here](https://github.com/PointCloudLibrary/pcl/compare/master...Ram-Z:modernize_cmake?expand=1#diff-af3b638bc2a3e6c650974192a53c7291R359).

However, this got a bit out of hand and I started modifying all pcl targets to provide their include and link libraries in their target interface. The more interesting changes are in [pcl_targets.cmake](https://github.com/PointCloudLibrary/pcl/compare/master...Ram-Z:modernize_cmake?expand=1#diff-b9e781cd471c6310c14c000be3a4ea0a).

Is this a change you would be interested in? I don't want to spend more time on this otherwise.

What minimum version of CMake do you want to support? It's currently set to 3.1, this was buildt using 3.14 and may use some newer features.

This builds all default libraries, tools were disabled for now.
```
-- The following subsystems will be built:
--   common
--   kdtree
--   octree
--   search
--   sample_consensus
--   filters
--   2d
--   geometry
--   io
--   features
--   ml
--   segmentation
--   visualization
--   surface
--   registration
--   keypoints
--   tracking
--   recognition
--   stereo
--   outofcore
--   people
-- The following subsystems will not be built:
--   apps: No reason
--   examples: Code examples are disabled by default.
--   simulation: Disabled by default.
--   global_tests: No reason
--   tools: Disabled manually.
```

taketwo commented:
We are certainly interested in modernizing our CMake scripts and welcome contributions in that direction.

What you decided to do is a great undertaking. We will need to be very careful though:
 * our CI pipelines neither test CMake scripts explicitly, nor package the library;
 * there are numerous corner cases and platform-dependent conditions in the scripts.

We will need extensive manual testing to make sure we don't break anything on any platform. I would prefer if we split this modernization over a series of as small as possible PRs which are easy to review and test.

> What minimum version of CMake do you want to support? 

Currently we require 3.1, however we may raise this to 3.5 if needed. Our current baseline environment is Ubuntu 16.04 (which has 3.5) and we'd prefer to not require anything higher than that unless absolutely needed.

Ram-Z commented:
Thanks for your reply, sorry it took so long to get back to you.

So far there was only one feature I used that required CMake>=3.5, which I have now reverted. I also managed to build the tools.

> I would prefer if we split this modernization over a series of as small as possible PRs which are easy to review and test.

I'm not entirely sure how this can be split up. I'm open to suggestions.

SergioRAgostinho commented:
Let's go progressively. My first suggestion would be to go first for component specification for vtk in `CMakeLists.txt`. It's a very small change which might have unforeseen consequences. I also see you
added the comment 
> #TODO this may not be sufficient if OpenGL2 is available

I believe that to be the case in the majority of platforms these days.  

I say we try that first and let it run through the CI, because right now you have no feedback whatsoever of what is building on osx and windows. It's a simply step but a necessary one.
